### PR TITLE
Make FLEXExplorerToolbar more extensible

### PR DESCRIPTION
**Pull request by NSExceptional**
Exposes `FLEXExplorerToolbar.toolbarItems` as a public API so that others can more easily, safely, and reliably add their own items to the FLEX toolbar.

### CODE REVIEW

File: Classes/Toolbar/FLEXExplorerToolbar.h [hunk 1]
```diff
@@ -12,6 +12,10 @@
 
 @interface FLEXExplorerToolbar : UIView
 
+/// The items to be displayed in the toolbar. Defaults to:
+/// globalsItem, hierarchyItem, selectItem, moveItem, closeItem
+@property (nonatomic, copy) NSArray *toolbarItems;
```
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.h]: Could you update the collection to be typed `<UIView>`?

File: Classes/Toolbar/FLEXExplorerToolbar.m [hunk 2]
```diff
@@ -150,9 +138,31 @@ - (void)layoutSubviews
     descriptionLabelFrame.size.width = CGRectGetMaxX(self.selectedViewDescriptionContainer.bounds) - kHorizontalPadding - descriptionOriginX;
     self.selectedViewDescriptionLabel.frame = descriptionLabelFrame;
 }
+    
+    
+#pragma mark - Setter Overrides
 
+- (void)setToolbarItems:(NSArray *)toolbarItems {
```
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Should we add a bail if `_toolbarItems == toolbarItems`?

File: Classes/Toolbar/FLEXExplorerToolbar.m [hunk 3]
```diff
@@ -150,9 +138,31 @@ - (void)layoutSubviews
...
-#pragma mark - Setter Overrides
+    for (FLEXToolbarItem *item in toolbarItems) {
+        [self addSubview:item];
+    }
+
+    _toolbarItems = toolbarItems.copy;
+
+    // Lay out new items
+    [self setNeedsLayout];
+    [self layoutIfNeeded];
```
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Let's remove this. It should be fine to let layout run without forcing it early.

**Comment by NSExceptional** [comment]
Good points! I'll take care of those! As for the last one though, I think it makes sense for the UI to update as soon as you set the property. This is how all UIKit views behave (`UITabBar.items`, `UINavigationItem.rightBarButtonItems`, etc). There's probably never a time one wouldn't want the UI to update right away.

Imo the cost of updating the UI prematurely is negligible, but if you want I could just set the ivar directly in the initializer and add the subviews there too

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 3 above)
Comment (NSExceptional) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: I think it makes sense for the UI to update as soon as you set the property. This is how all UIKit views behave (UITabBar.items, UINavigationItem.rightBarButtonItems, etc). There's probably never a time one wouldn't want the UI to update right away.

Imo the cost of updating the UI prematurely is negligible, but if you want I could just set the ivar directly in the initializer and add the subviews there too

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 2 above)
Comment (NSExceptional) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Done!

File: Classes/Toolbar/FLEXExplorerToolbar.h (hunk 1 above)
Comment (NSExceptional) [inline Classes/Toolbar/FLEXExplorerToolbar.h]: Done!

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 3 above)
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: I should clarify that the `setNeedsLayout` call is good and I only suggest removing the `layoutIfNeeded`. I poked into UITabBar out of curiosity, and didn't come across a `layoutIfNeeded` call in the surprisingly complicated path that comes from `setItems`.

**Comment by ryanolsonk** [comment]
Gonna go ahead and merge this. We can consider removing the `layoutIfNeeded` call but it's also not doing any real harm. Thanks!

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 3 above)
Comment (NSExceptional) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Ahhhhh gotcha, I see what you mean now. That could be removed, yeah

**Pull request merged**